tooling/nargo_cli/src/cli/trace_cmd.rs

```diff
@@ -36,9 +41,82 @@ pub(crate) struct TraceCommand {
     /// Directory where to store trace.json
     #[clap(long, short)]
     trace_dir: String,
+
+    /// Insert plonky2 information to the trace file
+    #[arg(long)]
+    trace_plonky2: bool,
+}
+
+fn generate_plonky2_debug_trace_list(
+    args: TraceCommand,
+    config: NargoConfig,
+) -> Result<DebugTraceList, CliError> {
+    let toml_path = get_package_manifest(&config.program_dir)?;
+    let selection = args.package.map_or(PackageSelection::DefaultOrAll, PackageSelection::Selected);
+    let workspace = resolve_workspace_from_toml(
+        &toml_path,
+        selection,
+        Some(NOIR_ARTIFACT_VERSION_STRING.to_string()),
+    )?;
+
+    let mut workspace_file_manager = file_manager_with_stdlib(std::path::Path::new(""));
+    insert_all_files_for_workspace_into_file_manager(&workspace, &mut workspace_file_manager);
+    let parsed_files = parse_all(&workspace_file_manager);
+
+    let Some(package) = workspace.into_iter().find(|p| p.is_binary()) else {
+        return Err(CliError::Generic(
+            "No matching binary packages found in workspace. Only binary packages can be debugged."
+                .to_string(),
+        ));
+    };
+
+    let compilation_result = compile_program(
+        &workspace_file_manager,
+        &parsed_files,
+        &workspace,
+        package,
+        &args.compile_options,
+        None,
+        true,
+        true,
+    );
+
+    // We need to get the debug trace list (if it exists), before calling report_errors(),
+    // because after that, the compilation_result variable gets moved and is no longer accessible.
+    let debug_trace_list = if let Ok((compiled_program, _)) = &compilation_result {
+        if let Some(plonky2_circuit) = &compiled_program.plonky2_circuit {
+            plonky2_circuit.debug_trace_list.clone()
+        } else {
+            // We don't return Err() here, because it doesn't explain to the user *why* the compilation failed
+            // to produce a plonky 2 circuit. Better call report_errors, which will give the user a more useful error list.
+            None
+        }
+    } else {
+        // Compilation result returned an Err? Again, better call report_errors, instead of failing here.
+        None
+    };
+
+    let _compiled_program = report_errors(
+        compilation_result,
+        &workspace_file_manager,
+        args.compile_options.deny_warnings,
+        args.compile_options.silence_warnings,
+    )?;
+
+    // Intentionally asserts that debug_trace_list has some value here, otherwise it's an internal compiler error,
+    // and the panic, caused by unwrap() is justified.
+    // If a debug_trace_list hasn't been produced, due to errors in the program,
+    // the previous call to report_errors should have returned an error value and we wouldn't have reached this point.
+    Ok(debug_trace_list.unwrap())
```

Comment:
Ok, I agree with the comments. Just a nit-picking suggestion: use expect, because if it's ever hit it would produce a string that can give more information and be used as a needle to search in the haystack of code:

```suggestion
    Ok(debug_trace_list.expect("report_errors did not produce error when it should have"))
```